contracts/gateway/OSTComposer.sol

```diff
@@ -0,0 +1,437 @@
+    );
+
+    /** Emitted whenever a revoke stake is called. */
+    event StakeRevoked(
+        address indexed staker,
+        bytes32 stakeRequestHash
+    );
+
+    /** Emitted whenever a reject stake is called. */
+    event StakeRejected(
+        address indexed staker,
+        bytes32 stakeRequestHash
+    );
+
+
+    /* Struct */
+
+    /**
+     * StakeRequest stores the stake amount, beneficiary address, gas price, gas
+     * limit, nonce, staker and gateway address.
+     */
+    struct StakeRequest {
+
+        /** Amount that will be staked. */
+        uint256 amount;
+
+        /**
+         * Address where the utility tokens will be minted in the
+         * auxiliary chain.
+         */
+        address beneficiary;
+
+        /** Gas price that staker is willing to pay. */
+        uint256 gasPrice;
+
+        /** Gas limit that staker is willing to pay. */
+        uint256 gasLimit;
+
+        /** Staker nonce at the gateway. */
+        uint256 nonce;
+
+        /** Address of the staker. */
+        address staker;
+
+        /** Address of the gateway where amount will be staked. */
+        EIP20GatewayInterface gateway;
+    }
+
+
+    /* Public Variables */
+
+    /* Mapping of staker to gateway to store stake request hash. */
+    mapping (address => mapping(address => bytes32)) public stakeRequestHashes;
+
+    /* Mapping of staker addresses to their StakerProxy. */
+    mapping (address => StakerProxy) public stakerProxies;
+
+    /* Stores number of all active stake request per staker. */
+    mapping(address => uint256) public activeStakeRequestCount;
+
+    /* Stores all the parameters of stake request based on stake request hash. */
+    mapping (bytes32 => StakeRequest) public stakeRequests;
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Contract constructor.
+     *
+     * @param _organization Address of an organization contract.
+     */
+    constructor(
+        OrganizationInterface _organization
+    )
+        Organized(_organization)
+        public
+    {
+
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Staker calls the method to show its intention of staking. In order
+     *         to stake, the staked amount must first be transferred to
+     *         OSTComposer. Staker should approve OSTComposer for token
+     *         transfer.
+     *
+     * @param _gateway Address of the gateway to stake.
+     * @param _amount Amount that is to be staked.
+     * @param _beneficiary The address in the auxiliary chain where the utility
+     *                     tokens will be minted.
+     * @param _gasPrice Gas price that staker is ready to pay to get the stake
+     *                  and mint process done.
+     * @param _gasLimit Gas limit that staker is ready to pay.
+     * @param _nonce Staker nonce specific to gateway.
+     *
+     * @return stakeRequestHash_ A unique hash for stake request.
+     */
+    function requestStake(
```

Comment:
This method should probably check that the gateway doesn't have any open stake request for the staker proxy. Otherwise, accept could fail.

Reply:
Ticket [OSTComposer optimization](https://github.com/OpenST/mosaic-chains/issues/79) to address above is created.